File: ecommerce-backend/app/routers/carts.py
import uuid

from fastapi import APIRouter, status

from app.db import SessionDep
from app.schemas import CartItemAdd, CartItemUpdate, CartRead, CartUpdate
from app.services.carts import CartService

router = APIRouter()


@router.post(
    "/carts",
    status_code=status.HTTP_201_CREATED,
)
async def create_cart(session: SessionDep):
    cart = await CartService.create_cart(session=session, items={})
    return str(cart.id)


@router.put(
    "/carts/{cart_id}",
    response_model=CartRead,
)
async def update_cart(
    cart_id: int,
    updated_cart: CartUpdate,
    session: SessionDep,
):
    return await CartService.update_cart(
        session=session, cart_id=cart_id, updated_cart=updated_cart
    )


@router.get("/carts/{cart_id}", response_model=CartRead)
async def get_cart(cart_id: int, session: SessionDep):
    return await CartService.get_cart(session=session, cart_id=cart_id)


@router.post("/carts/{cart_id}/items/", response_model=CartRead)
async def add_item_to_cart(
    cart_id: int,
    item: CartItemAdd,
    session: SessionDep,
):
    return await CartService.add_item_to_cart(
        session=session,
        cart_id=cart_id,
        product_id=str(item.product_id),
        quantity=item.quantity,
    )


@router.put("/carts/{cart_id}/items/{item_id}", response_model=CartRead)
async def update_item_in_cart(
    cart_id: int,
    item_id: uuid.UUID,
    item: CartItemUpdate,
    session: SessionDep,
):
    return await CartService.update_item_in_cart(
        session=session,
        cart_id=cart_id,
        product_id=str(item_id),
        quantity=item.quantity,
    )


@router.delete("/carts/{cart_id}/items/{item_id}", response_model=CartRead)
async def remove_item_from_cart(
    cart_id: int,
    item_id: uuid.UUID,
    session: SessionDep,
):
    return await CartService.remove_item_from_cart(
        session=session,
        cart_id=cart_id,
        product_id=str(item_id),
    )

File: ecommerce-storefront/src/services/carts.ts
import { fetchApi } from './api';

interface StoreError {
  message: string;
}

export interface CartItem {
  product_id: string;
  name: string;
  description: string;
  image_url: string;
  quantity: number;
  price: number;
  line_total: number;
  stock: number;
}

export interface CartSummary {
  subtotal: string;
  grand_total: string;
  total_items_count: number;
}

export interface Cart {
  id: number;
  user_id: string | null;
  items: CartItem[];
  summary: CartSummary;
  updated_at: string;
}

export async function createCart(): Promise<number> {
  const res = await fetchApi('/carts', {
    method: 'POST',
  });
  if (!res.ok) throw new Error('Failed to create cart');
  // The backend returns the ID as a string in the response body
  const cartId = await res.json();
  return parseInt(cartId);
}

export async function getCart(cartId: number): Promise<Cart> {
  const res = await fetchApi(`/carts/${cartId}`);
  if (!res.ok) throw new Error('Failed to fetch cart');
  return res.json();
}

export async function addItemToCart(
  cartId: number,
  productId: string,
  quantity: number
): Promise<Cart | StoreError> {
  const res = await fetchApi(`/carts/${cartId}/items/`, {
    method: 'POST',
    body: JSON.stringify({ product_id: productId, quantity }),
  });
  if (!res.ok) {
    const error = await res.json();
    console.log(error);
    if (res.status === 400 || res.status === 404) {
      return { message: error.detail };
    }
    return { message: 'Failed to add item to cart' };
  }
  return await res.json();
}

export async function updateItemInCart(
  cartId: number,
  productId: string,
  quantity: number
): Promise<Cart> {
  const res = await fetchApi(`/carts/${cartId}/items/${productId}`, {
    method: 'PUT',
    body: JSON.stringify({ quantity }),
  });
  if (!res.ok) throw new Error('Failed to update item in cart');
  return res.json();
}

export async function removeItemFromCart(
  cartId: number,
  productId: string
): Promise<Cart> {
  const res = await fetchApi(`/carts/${cartId}/items/${productId}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error('Failed to remove item from cart');
  return res.json();
}
